File: src/main/java/me/lokka30/levelledmobs/rules/ChunkKillOptions.java
package me.lokka30.levelledmobs.rules;

import org.jetbrains.annotations.Nullable;

public class ChunkKillOptions implements MergableRule, Cloneable{
    public Boolean disableVanillaDrops;
    public Boolean disableItemBoost;
    public Boolean disableXpDrops;

    public ChunkKillOptions(){
        disableItemBoost = true;
        disableXpDrops = true;
    }

    public boolean isDefault(){
        return (disableVanillaDrops == null && disableItemBoost == null && disableXpDrops == null);
    }

    public boolean getDisableVanillaDrops(){
        return this.disableVanillaDrops != null && this.disableVanillaDrops;
    }

    public boolean getDisableItemBoost(){
        return this.disableItemBoost != null && this.disableItemBoost;
    }

    public boolean getDisableXpDrops(){
        return this.disableXpDrops != null && this.disableXpDrops;
    }

    public void merge(final @Nullable MergableRule mergableRule){
        if (!(mergableRule instanceof final ChunkKillOptions chunkKillOptions)){
            return;
        }

        if (chunkKillOptions.isDefault()) return;

        if (chunkKillOptions.disableVanillaDrops != null)
            this.disableVanillaDrops = chunkKillOptions.disableVanillaDrops;
        if (chunkKillOptions.disableItemBoost != null)
            this.disableItemBoost = chunkKillOptions.disableItemBoost;
        if (chunkKillOptions.disableXpDrops != null)
            this.disableXpDrops = chunkKillOptions.disableXpDrops;
    }

    public boolean doMerge(){
        return true;
    }

    public ChunkKillOptions cloneItem() {
        ChunkKillOptions copy = null;
        try {
            copy = (ChunkKillOptions) super.clone();
        } catch (final Exception e) {
            e.printStackTrace();
        }

        return copy;
    }

    public String toString(){
        if (this.isDefault()) return "Default";

        final StringBuilder sb = new StringBuilder();
        if (disableVanillaDrops != null && disableVanillaDrops) {
            sb.append("disableVanillaDrops");
        }
        if (disableItemBoost != null && disableItemBoost) {
            if (!sb.isEmpty()) sb.append(", ");
            sb.append("disableItemBoost");
        }
        if (disableXpDrops != null && disableXpDrops) {
            if (!sb.isEmpty()) sb.append(", ");
            sb.append("disableXpDrops");
        }

        return sb.toString();
    }
}

File: levelledmobs-plugin/src/main/kotlin/io/github/arcaneplugins/levelledmobs/rules/ChunkKillOptions.kt
package io.github.arcaneplugins.levelledmobs.rules

/**
 * Holds various options relating to the
 * chunk kill feature
 *
 * @author stumper66
 * @since 3.12.0
 */
class ChunkKillOptions : MergableRule, Cloneable {
    var disableVanillaDrops: Boolean? = null
    var disableItemBoost: Boolean? = true
    var disableXpDrops: Boolean? = true

    val isDefault: Boolean
        get() = (disableVanillaDrops == null && disableItemBoost == null && disableXpDrops == null)

    fun getDisableVanillaDrops(): Boolean {
        return this.disableVanillaDrops != null && disableVanillaDrops!!
    }

    fun getDisableItemBoost(): Boolean {
        return this.disableItemBoost != null && disableItemBoost!!
    }

    fun getDisableXpDrops(): Boolean {
        return this.disableXpDrops != null && disableXpDrops!!
    }

    override fun merge(mergableRule: MergableRule?) {
        if (mergableRule !is ChunkKillOptions) return
        if (mergableRule.isDefault) return

        if (disableVanillaDrops != null) this.disableVanillaDrops = mergableRule.disableVanillaDrops
        if (disableItemBoost != null) this.disableItemBoost = mergableRule.disableItemBoost
        if (disableXpDrops != null) this.disableXpDrops = mergableRule.disableXpDrops
    }

    override val doMerge = true

    override fun cloneItem(): Any {
        var copy: ChunkKillOptions? = null
        try {
            copy = super.clone() as ChunkKillOptions
        } catch (e: Exception) {
            e.printStackTrace()
        }

        return copy as Any
    }

    override fun toString(): String {
        if (this.isDefault) return "Default"

        val sb = StringBuilder()
        if (disableVanillaDrops != null) {
            if (getDisableVanillaDrops()) sb.append("disableVanillaDrops")
            else sb.append("disableVanillaDrops: false")
        }
        if (disableItemBoost != null) {
            if (sb.isNotEmpty()) sb.append(", ")
            if (getDisableItemBoost()) sb.append("disableItemBoost")
            else sb.append("disableItemBoost: false")
        }
        if (disableXpDrops != null) {
            if (sb.isNotEmpty()) sb.append(", ")
            if (getDisableXpDrops()) sb.append("disableXpDrops")
            else sb.append("disableXpDrops: false")
        }

        return sb.toString()
    }
}